### DISCUSSION TITLE
Open app with NFC tag and pass routing information

### DISCUSSION THREAD

**Discussion by NePheus** [ionic-team/capacitor#4411] [Q&A] [answered]
Hello,

my goal is to open my app with a specific route when I hold it over a NFC tag. So far I tried it with deep linking. Deep linking itself is working as expected. When I click the deep link, my app gets opened automatically and is routing correctly to the destination page. When I write this URL with the Android App "NFC Tools" on my NFC tag and hold my phone on it, it just opens Chrome but not the app.
I read on this page, that deep linking is blocked with NFC tags:
https://joturl.zendesk.com/hc/en-us/articles/360014154460-How-to-deep-link-into-your-Android-app-from-an-NFC-tag
"_Let's start by saying that Android blocks normal deep link techniques if the URL comes from an NFC tag. It is a security protocol and there is not much you can do to counter it._"
Is this right?

Is there another way to achieve this, i.e. using intents? If yes, is there any tutorial about this?

**Comment by NePheus** [comment]
If I change this:
```
<intent-filter android:autoVerify="true">
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.DEFAULT" />
    <category android:name="android.intent.category.BROWSABLE" />
    <data android:host="@string/deep_link_url" android:scheme="https" />
</intent-filter>
```
to this:
```
<intent-filter android:autoVerify="true">
    <action android:name="android.nfc.action.NDEF_DISCOVERED"/>
    <category android:name="android.intent.category.DEFAULT" />
    <data android:host="@string/deep_link_url" android:scheme="https" />
</intent-filter>
```
The app starts with the nfc tag, but no listener is triggering, like "App.addListener('appUrlOpen'..." or "App.addListener('appRestoredResult'...", so I cant route to my destination.

**Reply by jcesarmobile** [reply]
App plugin responds to `VIEW` action, you'll need a plugin to respond to `NDEF_DISCOVERED` action

**Reply by NePheus** [reply]
Ok I understand. I gave this a try:
https://ionicframework.com/docs/native/nfc
But I get a NullPointerException caused by com.chariotsolutions.nfc.plugin.NfcPlugin.sendEvent when the app starts after holding the phone on the NFC tag. Maybe because I am using Capacitor 3. So I found another fork here:
https://github.com/iotize-sas/device-com-nfc.cordova
This is running without this error.
So I have the following setup now:
AndroidManifest.xml
```
<intent-filter android:autoVerify="true">
    <action android:name="android.nfc.action.NDEF_DISCOVERED"/>
    <category android:name="android.intent.category.DEFAULT" />
    <data android:mimeType="application/MYBUNDLEID" />
</intent-filter>
```
app component constructor:
```
this._nfc
  .addNdefListener(
    function () {
      console.log("addNdefListener success");
    },
    function (reason) {
      alert("addNdefListener failed " + reason);
    }
  )
  .subscribe((nfcEvent) => {
    var tag = nfcEvent.tag,
      ndefMessage = tag.ndefMessage;
    alert(this._nfc.bytesToString(ndefMessage[0].payload).substring(1));
  });

this._nfc
  .addMimeTypeListener(
    "application/MYBUNDLEID",
    function () {
      console.log("addNdefListener success");
    },
    function (reason) {
      alert("addNdefListener failed " + reason);
    }
  )
  .subscribe((nfcEvent) => {
    var tag = nfcEvent.tag,
      ndefMessage = tag.ndefMessage;
    alert(this._nfc.bytesToString(ndefMessage[0].payload).substring(1));
  });
```
Like described here:
https://github.com/don/nfc-launch/blob/master/www/js/index.js

Result:
I get my alert when the app is started and I hold my phone on the NFC tag. When I close the app and hold it on the NFC tag, the app starts, but I do not get the dialog.
I think this thread is about that issue https://github.com/chariotsolutions/phonegap-nfc/issues/187 but I dont know how to handle it with capacitor 3.

**Comment by NePheus** [comment, marked as answer]
I created a capacitor 3 plugin for app launching with nfc:
https://www.npmjs.com/package/capacitor-nfc-launch
